### ISSUE TITLE
disconnected on invalid subscribe topic

### ISSUE THREAD

**Issue by dustin** [eclipse/mosquitto#1413]
When trying to subscribe to an invalid topic, the broker without reason.

This seems inconsistent with:

> When the Server receives a SUBSCRIBE packet from a Client, the Server MUST respond with a SUBACK packet [MQTT-3.8.4-1].

<img width="601" alt="Screen Shot 2019-09-15 at 22 26 16" src="https://user-images.githubusercontent.com/1779/64935782-dae01500-d807-11e9-85af-67bb1f7c2231.png">

[badsub.pcap.gz](https://github.com/eclipse/mosquitto/files/3614811/badsub.pcap.gz)

**Comment by cjmaynard** [comment]
Certainly Roger and the other mosquitto experts can weigh in here, but I think this is being treated as a protocol violation, and is therefore handled properly according to Section 4.8 "Handling errors", where it states the following:

`Unless stated otherwise, if either the Server or Client encounters a protocol violation, it MUST close the Network Connection on which it received that Control Packet which caused the protocol violation`

Ref: http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.pdf

**Comment by dustin** [comment]
IMO because the packet is well-formed and the protocol provides an error channel that would be appropriate to use, the broker should report an error.

In 3.1.1, the broker only has the ability to NAK my subscription attempt for a specific topic, which is still pretty useful for a client user.

In 5.0, the topic can be NAK'd with a failure type and a descriptive reason.

**Comment by cjmaynard** [comment]
Perhaps, but within 4.7.1 it states, "The wildcard characters can be used in Topic Filters, but MUST NOT be used within a Topic Name" and within 4.7.1.2 it states, "The multi-level wildcard character MUST be specified either on its own or following a topic level separator. In either case it MUST be the last character specified in the Topic Filter", so this appears to be a clear protocol violation, does it not?

That said, whether it's reasonable or not to change the behavior so that a SUBACK with failure (0x80) return code is delivered instead of closing the connection, I leave to the mosquitto developers to decide.

**Comment by dustin** [comment]
> so this appears to be a clear protocol violation, does it not?

I don't think it's a protocol violation because it's a different layer.  The topic is effectively user input (I found it in a program that was using my client).

The client *could* attempt to validate topic names, as `mosquitto_sub` seems to do, but this seems unnecessary since it requires every client to validate data that it could otherwise pass as an opaque value and that the broker (which is already deciding how to process the topic) can decide it's invalid and send a consistent message back to the client which can, in turn, present something meaningful to the user.

The current result is that an application user asking for something the server considers invalid just gets a disconnect error with no reason, and, as a user, I'd think this would make things confusing.

(again, I can add code to prevent this, predicting when the hangup might occur and synthesizing errors, but that seems redundant and incomplete).

I understand what you're saying, and I don't think you're entirely wrong, but I do think the specification has an excellent channel for handling this gracefully and with good communication and using it here would be a large improvement.

**Comment by karlp** [comment]
I would agree with cjmaynard, this is an illegal topic.  your client library (if in use) should be rejecting this already, if you send it to the broker, I would think it completely within the understanding of the spec to disconnect rather than nak.

**Comment by dustin** [comment]
The arguments here seem to be along the lines of "the server does this and it's allowed to."  The server is not *required* to disconnect on a bad topic and I don't see why anyone would argue that it's a good idea to *require* client side validation of data that's already being validated, presumably correctly, on the server.

I understand that it may not be a spec violation to hang up if you receive something you don't like.  I don't see how it's a good idea, though.

The broker is validating the input and has a channel for reporting the status of that validation.  There could be tons of state that's getting thrown away rather than just reporting an error using this channel.

It certainly is possible for every client implementation to anticipate every error that doesn't pass every broker's validation, but that seems quite unnecessary.  Even if I get it right, the server itself may have a bug and decide to hang up on me because it perceived a particular topic invalid and I'm left to guess what the server didn't like among all the things I sent.

Protocols with error channels typically use their error channels to communicate messages with understood, but invalid contents.  We have that mechanism.  Using it will enhance stability of applications and help people understand when they make mistakes.

My particular use case that led to this was attempting to subscribe to an invalid topic so I could verify the error path communicated clearly to my users.  I figured this was an easy way to get the server to send one of the subscribe errors.  At the moment, I don't know when it might send one and when it might just hang up on me.

**Comment by ralight** [comment]
Mosquitto has been using DISCONNECT with a reason code for v5.0 clients doing this since 2.0, so I'm closing this issue
